### DISCUSSION TITLE
Connect Bluetooth Laser Tag gun

### DISCUSSION THREAD

**Discussion by robegamesios** [LucidVR/opengloves-driver#206] [Ideas]
hi, im wondering if its possible to mod this driver to support bluetooth laser tag gun like the Recoil RK 45 https://www.youtube.com/watch?v=S81l8VBC7dw ? it seems a good candidate since it has electric blowback. Maybe the phone holder can be used to hold the oculus quest 2 touch controller for tracking. then the trigger, comm button, reload button can be mapped for the game. I wanna make it work primarily for half life Alyx. and maybe use it for other games in the future. thanks

**Comment by LittleNyanCat** [comment]
Should be doable as long as you can figure out how to talk with it

**Reply by robegamesios** [reply]
I ordered the esp-32 board (waiting for it), might be easier to swap it with the bt module that came with the gun.  im also still waiting for the gun to arrive but from what I've seen on videos, the electric blowback (ebb) activates on the press of the trigger and is not dependent on the bt comm. If I can map the buttons of the oculus touch controller to dedicated buttons in the gun, then it might just work. The biggest question is probably when the gun magazine goes empty, how to prevent the ebb from activating.

**Reply by LittleNyanCat** [reply]
You'll want connect the buttons to the esp that part is trivial. For the disabling of the ebb, can do it with a transistor controlled by the esp , and then you'll need a mod in your game to tell the esp when to enable or disable ebb

**Reply by robegamesios** [reply]
would love some context on how to disable the ebb and mod the game. I'll try to just make it work as a start, I'll post any progress I make. feel free to comment if you have any suggestions. thanks!

**Reply by LittleNyanCat** [reply]
The ebb mechanism in that gun you sent the link of seems to be driven by a simple DC motor, meaning you could ignore the main board entirely and drive it with a MOSFET like this: (Ignore that the image has an Arduino Uno, it works the same way for the ESP32 as well)

![Alt text](http://adam-meyer.com/arduino/images/2012/03/rfp30n06le-arduino-motor.png "Mosfet circuit")

As for the mod part of the game I don't know enough to help you a lot on that part, but I do know that you need to add a way of the game telling opengloves when to enable ebb, so opengloves can pass on that information to the esp and then have the firmware running on it interpret that information to enable or disable the pin controlling the MOSFET. You might want to take a look at the already existing force feedback mod and change it so it includes when to enable ebb based on game events.

By the way - you might want to join our discord, although your project isn't exactly a glove, it'll still rely heavily on the opengloves driver and there are peeps in there that are much knowledgeable than me in the software and electronics that can help you out.

**Reply by robegamesios** [reply]
oh thanks for the great info. im not familiar with MOSFET (I'll do some research) but on high level, why is it needed?  

Im in the discord channel. I see mine and your comments get posted by a git bot there :) Yep great crowd in the discord that I'm sure can help, hopefully others are interested too

**Reply by LittleNyanCat** [reply]
> oh thanks for the great info. im not familiar with MOSFET (I'll do some research) but on high level, why is it needed?

The pins of the ESP can't handle the current the motor needs directly, so the MOSFET acts as an electrical switch controlled by the pin which can handle the current

**Reply by robegamesios** [reply]
Ok,  can I use the parts from this post? https://diyi0t.com/control-dc-motor-without-ic-motor-driver/ , seems like a match for the gun motor, the gun uses 4 AA batteries. And the diode and resistors are also needed?

**Reply by LittleNyanCat** [reply]
Yes it should work, and yes you do need those extra components, although I've used a MOSFET without that 220 ohm resistor between it and the pin before. The other resistor and the diode are strictly required.

**Reply by robegamesios** [reply]
ok cool I just ordered them. so this setup will power the ebb via the esp32 right? so for this diagram, which wire will connect to the gun trigger?
<img width="1066" alt="Screen Shot 2022-06-28 at 2 32 38 PM" src="https://user-images.githubusercontent.com/10041871/176302101-8b40570f-08fd-4fde-85b2-c1d16620e19c.png">

**Reply by LittleNyanCat** [reply]
>this setup will power the ebb via the esp32 right? 

yeah it will, which might not be the best idea (at least not from the 3.3v pin), as the motor on the gun looks a little beefier, but that's easy to fix. 

for the trigger one of the wires of the trigger will connect to gnd and the other goes to one of the pins of the esp (by default, 12)

**Reply by LittleNyanCat** [reply]
and the same goes for the other buttons - one wire to gnd another to a pin on the esp

**Reply by robegamesios** [reply]
> 
> yeah it will, which might not be the best idea (at least not from the 3.3v pin), as the motor on the gun looks a little beefier, but that's easy to fix.
> 
do I have to use the 3.3v pin or can I use the 5V pin? if not, what's the most likely fix? add a different power source for the motor?

**Reply by LittleNyanCat** [reply]
>can I use the 5V pin? if not, what's the most likely fix? add a different power source for the motor?

yes and yes. using the 5v pin will essentially connect the motor to the ESP's microusb port (through a protection diode), but the motor will be weaker that way compared to powering it with 4 AAs like it was originally meant to (which would give you around 6v).  fortunately you can power the motor from AAs (or another source of your choosing) by connecting the gnd of your power source to the gnd of your esp and vcc to the motor's vcc

**Reply by robegamesios** [reply]
I am planning to use a power bank as power source via the micro USB port. I have one of those slim candy bar ones that outputs 1A max 5V that im thinking I could mount in the gun. but if it does not suffice, then I'll have to try the bigger power banks and just put it in my pocket.  

`fortunately you can power the motor from AAs (or another source of your choosing) by connecting the gnd of your power source to the gnd of your esp and vcc to the motor's vcc`

not sure how to power the esp using the AAs like you mentioned above without seeing a visual diagram (sorry). but I'll try the power bank first and see if it can do the job.

here is the esp32 I ordered from amazon as suggested: https://www.amazon.com/dp/B07QCP2451?psc=1&ref=ppx_yo2ov_dt_b_product_details

<img width="1143" alt="Screen Shot 2022-06-28 at 4 09 37 PM" src="https://user-images.githubusercontent.com/10041871/176320421-22a4f846-b8b7-4f36-92a8-01c37f80220e.png">

**Reply by robegamesios** [reply]
so the diagram above, I'll connect the Vin 5V going to the motor right?

**Reply by LittleNyanCat** [reply]
yeah, connect it to 5v instead of 3.3v

**Reply by robegamesios** [reply]
ok great! I'll post update once I get the parts and start prototyping. thanks!!

**Reply by robegamesios** [reply]
`You might want to take a look at the already existing force feedback mod and change it so it includes when to enable ebb based on game events.`

 @LittleNyanCat this makes a lot of sense. Now that I think about it, maybe the motor should be connected to a haptic feedback signal from the game to esp32, rather than connected to the trigger button. so the flow would go like:

trigger pressed and sends signal to game -> if gun fired (not sure how this is sent for haptic feedback - maybe sound?) then send signal to esp32 that will activate the ebb. 

This way the ebb only fires via the gun-fired-event. Or if different events can be sent to esp32, then maybe the ebb motor can be set to different speeds to differentiate single-bullet shot, triple-bullet shot, maybe even the Alyx multi-tool. Probably also cover the use case if empty magazine, then don't trigger ebb.

is the repo for the `open gloves half life Alyx integration` public? Im having trouble finding it. Im guessing modding it would accomplish what I want? 

@danwillm @lucas-vrtech would love your input on this if it's possible. thanks

**Reply by robegamesios** [reply]
I've figured out how to listen to the `player_shoot_weapon` game event in HLA. I modified the openGloves.lua to include this game listener event. I am having trouble trying to run force feedback mod though https://github.com/LucidVR/opengloves-half-life-alyx-scripts. I see these screens below when launching the force feedback integration. 

Am I suppose to follow what this dialog box is telling me?
![IMG_9716](https://user-images.githubusercontent.com/10041871/177870235-f5ac51a1-f750-40f4-b764-fb64ab433d78.jpg)

I clicked the `Half-Life: Alyx Integration` button and this is what pops up. Is the game suppose to launch automatically (it's stuck here and does not launch anything) OR do I need to launch HLA game separately? When I try to launch HLA from steam, this force feedback integration app exits (I checked task manager and it's no longer running).
![IMG_9715](https://user-images.githubusercontent.com/10041871/177870300-81f2e3ee-bc54-40ba-bd9f-8044a5b9799c.jpg)

How does the openGloves.lua file communicate with the driver to the esp32? Looking at the lua file https://github.com/LucidVR/opengloves-half-life-alyx-scripts/blob/main/game/hlvr/scripts/vscripts/opengloves.lua

function    outputData(attachment)
  line = table.concat(fingerPos[attachment], ", ")
  
  if attachment == LEFT_ID then
    Msg("[OpenGlovesParse]{Left} ("..line..")\n")
  elseif attachment == RIGHT_ID then
    Msg("[OpenGlovesParse]{Right} ("..line..")\n")
  else
    Msg("[OpenGlovesError] Invalid attachment \n")
  end
  --local file = io.open("openglovesdata.json", "w")
  --io.output(file)
  --io.write(line)
  --io.close(file)
end

seems this is suppose to the communicating? but the last lines are commented? I need to send the signal from the gameEvent to the esp32 to drive the motor. 

thanks again
